### ISSUE TITLE
Loss NaN Issue During Training with Empiar10180 Data

### ISSUE THREAD

**Issue by kamzero** [alncat/opusDSD#11]
**Describe the bug**
During the training process of the model using the Empiar10180 dataset, I encounter NaN (Not a Number) values in the loss approximately between 30,000 to 60,000 iterations. This issue is inconsistent with the expected numerical outputs and disrupts the training procedure. Here is my terminal output and log:

> Additional info at 60480, l1=0.00152, tv=0.00093, snr=0.0101, mse=0.3467, gen_loss=-0.00346, bc=0.0088                 Additional info at 61440, l1=0.00152, tv=0.00087, snr=0.0101, mse=0.3463, gen_loss=-0.00347, bc=0.0088                 Additional info at 62400, l1=0.00152, tv=0.00088, snr=0.0103, mse=0.3467, gen_loss=-0.00352, bc=0.0090                 Additional info at 63360, l1=0.00151, tv=0.00092, snr=0.0101, mse=0.3468, gen_loss=-0.00349, bc=0.0088                 Additional info at 64320, l1=0.00152, tv=0.00092, snr=0.0104, mse=0.3464, gen_loss=-0.00355, bc=0.0091                 Additional info at 65280, l1=0.00152, tv=0.00090, snr=0.0104, mse=0.3465, gen_loss=-0.00358, bc=0.0091                 Additional info at 66240, l1=0.00151, tv=0.00079, snr=0.0103, mse=0.3466, gen_loss=-0.00354, bc=0.0090                 Additional info at 67200, l1=nan, tv=nan, snr=nan, mse=0.3105, gen_loss=-0.00317, bc=nan                               Additional info at 68160, l1=nan, tv=nan, snr=nan, mse=0.2296, gen_loss=-0.00234, bc=nan                               Additional info at 69120, l1=nan, tv=nan, snr=nan, mse=0.1699, gen_loss=-0.00173, bc=nan                               Additional info at 70080, l1=nan, tv=nan, snr=nan, mse=0.1257, gen_loss=-0.00128, bc=nan                               Additional info at 71040, l1=nan, tv=nan, snr=nan, mse=0.0929, gen_loss=-0.00095, bc=nan                               Additional info at 72000, l1=nan, tv=nan, snr=nan, mse=0.0688, gen_loss=-0.00070, bc=nan                               Additional info at 72960, l1=nan, tv=nan, snr=nan, mse=0.0509, gen_loss=-0.00052, bc=nan                               Additional info at 73920, l1=nan, tv=nan, snr=nan, mse=0.0376, gen_loss=-0.00038, bc=nan                               Additional info at 74880, l1=nan, tv=nan, snr=nan, mse=0.0278, gen_loss=-0.00028, bc=nan      
             
[run.log](https://github.com/alncat/opusDSD/files/14586861/run.log)

**To Reproduce**
My command is: 
`dsd train_cv all.mrcs --ctf ./sp-ctf.pkl --poses ./sp-pose-euler.pkl --lazy-single --pe-type vanilla --encode-mode grad --template-type conv -n 20 -b 4 --zdim 12 --lr 1.e-5 --num-gpus 8 --multigpu --beta-control 2. --beta cos -o /mnt/nas/share2/home/jxr/code/opusDSD/work_dirs/10180/sp_b4_g8_lr-5 -r ./mask.mrc --downfrac 0.75 --valfrac 0.25 --lamb 1. --split sp-split.pkl --bfactor 4. --templateres 224`

**Expected behavior**
I anticipated that the training would continue without any numerical errors.

**Additional context**
Due to memory constraints (I have a 24GB GPU with an RTX 4090), I adjusted the batch size to be smaller (12->4). I suspect that the reduction in batch size may have led to training instability. As a result, I increased the number of GPUs (from 4 to 8) and decreased the learning rate (from 1e-4 to 1e-5), but there was no significant improvement.

I would like to ask if you have any suggestions regarding this issue? I would be very grateful.

**Comment by alncat** [comment]
Hi, it looks like the gen_loss is too low. For this dataset, the expected loss is around -0.02. You may checkout this issue, https://github.com/alncat/opusDSD/issues/10. The configuration of environment might be the cause of NAN.

**Comment by alncat** [comment]
I have created a smaller test set In this issue. You may follow the discussion here. You can also include a --plot argument during training so dsd will report the correlations between reconstructions and training data, and output temporary reconstructions like tmp*.mrc in the training directory.

**Comment by kamzero** [comment]
Hello,

Thank you for your guidance. I have run the small test case as suggested in issue #10 and I'm glad to report that there were no NaN issues encountered. The log is as follows:
[run.log](https://github.com/alncat/opusDSD/files/14597180/run.log)

My environment is CUDA 11.3 with PyTorch 1.11.0. It seems the issue may not be related to the environment after all.

**Comment by alncat** [comment]
This run looks OK~ Did you check the reconstructions? Do they look like the spliceosome? And can you check the order of images in all.star and the order of images in the starfile you used to create pose-euler.pkl? Are they in the same order?

**Comment by kamzero** [comment]
Thank you for your suggestion~ I believe the reconstruction results from the small test case appear to be normal and resemble the structure of a spliceosome. 
![image](https://github.com/alncat/opusDSD/assets/40420631/3ccea37d-2f76-48d6-91d8-e78267a0b85b)

I am not entirely clear on how to rigorously check the order of images in the .star file. I have simply opened it in text format, and it seems that the _rlnImageName in the all.star file is largely consistent with the original .star file used to create the pose-euler.pkl. ![image](https://github.com/alncat/opusDSD/assets/40420631/880eac40-4058-46b0-86cb-0ba0c297c686)
I have attempted to create pose-euler.pkl from the all.star file, but I still encountered NaN values.
Do you have any other suggestions? I would be very grateful :D

**Comment by alncat** [comment]
Hi~ You can try another smaller test dataset, https://empiar.pdbj.org/entry/10492. This one only contains 50k images. If you can obtain expected results on this set, then the program installation should be good. I will upload it to a web drive and share with you.

**Comment by alncat** [comment]
Hi @kamzero , you can try files in this link: https://pan.baidu.com/s/1c7UVHYwzZTjsbLQUZY6gPw?pwd=6xgq extraction code: 6xgq

**Comment by kamzero** [comment]
Thanks very much, I will have a try~

**Comment by alncat** [comment]
You are welcome, hope it works 😄

**Comment by kamzero** [comment]
> You are welcome, hope it works 😄

It seems all right, thanks for your time and consideration :D
